### DISCUSSION TITLE
Class emulating numeric behaviour with Pint quantities

### DISCUSSION THREAD

**Discussion by nopria** [hgrecco/pint#1903] [Q&A] [answered]
I wrote a custom Point class which I'd like to work with Pint quantities but I get a strange error if I try to emulate numeric behaviour:

```
import pint
um = pint.get_application_registry()

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y
    def __add__(self, other):
        if isinstance(other, pint.Quantity):
            return Point(self.x + other, self.y + other)
        return NotImplemented
    def __radd__(self, other):
        if isinstance(other, pint.Quantity):
            return Point(self.x + other, self.y + other)
        return NotImplemented

q = 0.03 * um.m
p = Point(10 * um.mm, 2 * um.cm)
```

`p + q`  works correctly as expected, while `q + p` gives the followong error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\UP\miniconda3\lib\site-packages\pint\facets\plain\quantity.py", line 841, in __add__
    return self._add_sub(other, operator.add)
  File "C:\Users\UP\miniconda3\lib\site-packages\pint\facets\plain\quantity.py", line 103, in wrapped
    return f(self, *args, **kwargs)
  File "C:\Users\UP\miniconda3\lib\site-packages\pint\facets\plain\quantity.py", line 739, in _add_sub
    raise DimensionalityError(self._units, "dimensionless")
pint.errors.DimensionalityError: Cannot convert from 'meter' to 'dimensionless'
```

Any clue? Thanks in advance.

**Comment by andrewgsavage** [comment]
add Point to pint.compat.upcast_type_map

see https://github.com/hgrecco/pint/issues/1888

**Comment by nopria** [comment]
I did

`pint.compat.upcast_type_map.update({'Point':None})
`

but nothing changes.

**Comment by andrewgsavage** [comment]
try 
`pint.compat.upcast_type_map.update({'__main__.Point':None}) `

**Reply by hgrecco** [reply]
Shall we probide a `register_upcast_type` function or similar that does `the right thing` in relation to naming?

**Comment by nopria** [comment]
In this specific MWE the last above solution works, but in my actual code Pint is not imported in the same script where class Point is defined but in a separate module `tools` which is imported by `import tools as tls`. With such configuration just adding

    pint.compat.upcast_type_map.update({'__main__.Point':None})

to `tools.py` does not work. Moreover, `Point` is a subclass of `Geotools`, maybe this affects the instruction needed to solve my issue?

**Reply by andrewgsavage** [reply, marked as answer]
Run type(object) and it will give the string you should use. For the example above I did
```
type(p)
__main__.Point
```

**Reply by nopria** [reply]
Now it works, thanks.
